presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -79,31 +90,163 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
             return Optional.of(new DecorrelationResult(
                     node,
                     ImmutableSet.of(),
-                    ImmutableList.of()));
+                    ImmutableList.of(),
+                    ImmutableMultimap.of(),
+                    false));
         }
 
         @Override
         public Optional<DecorrelationResult> visitFilter(FilterNode node, Void context)
         {
-            Expression predicate = node.getPredicate();
-            if (!SymbolsExtractor.extractUnique(predicate).containsAll(correlation)) {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = Optional.of(new DecorrelationResult(
+                    node.getSource(),
+                    ImmutableSet.of(),
+                    ImmutableList.of(),
+                    ImmutableMultimap.of(),
+                    false));
+
+            // try to decorrelate filters down the tree
+            if (containsCorrelation(node.getSource(), correlation)) {
```

Comment:
The invariant for visiting child is: "visit child only when it has correlated symbols", because otherwise there is nothing to decorrelate.

Exception is this TODO (to be fixed): https://github.com/prestodb/presto/pull/10204/files#diff-19133f604c15d7a69f1fa60c49cc394bR67
Also you can also skip "child has correlation" for other nodes than `FilterNode` because they won't be able to de-correlate without information from `FilterNode`.